### PR TITLE
UIGestureRecognizer trigger

### PR DESCRIPTION
**Pull request by bitomule**
I've added a new Reactive extension for UIGestureRecognizer with just a trigger method. Creates a signal from a reactive UIGestureRecognizer:

```
let tapViewGesture = UITapGestureRecognizer()
fieldTapGestureRecognizer.reactive.trigger().observeValues({ _ in

})
```

### CODE REVIEW

**Comment by bitomule** [comment]
Based on https://github.com/ReactiveCocoa/ReactiveCocoa/pull/3256

**Comment by andersio** [comment]
If there is no argument, a computed variable would be fine.

File: ReactiveCocoa/UIKit/UIGestureRecognizer.swift [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import enum Result.NoError
+
+extension Reactive where Base: UIGestureRecognizer {
+	
+	/// Create a signal which sends a `next` event for each gesture event
+	///
+	/// - returns:
+	///   A trigger signal.
+	
+	public func trigger() -> Signal<(), NoError> {
```
Comment (andersio) [inline ReactiveCocoa/UIKit/UIGestureRecognizer.swift]: If there is no argument, a computed property would be fine.

Comment (bitomule) [inline ReactiveCocoa/UIKit/UIGestureRecognizer.swift]: recognised seems fine to me. I'll do the changes based on your feedback.

**Comment by bitomule** [comment]
I've changed the name to recognised and now it's a computed property.

**Comment by andersio** [comment]
Would you mind to write a unit test for it?

**Comment by bitomule** [comment]
Of course. Is there any example I can follow?

**Comment by andersio** [comment]
You can just follow those in `ReactiveCocoaTests/UIKit/`.

**Comment by bitomule** [comment]
I wasn't sure on how to test a UIGestureRecognizer, finally found this (https://github.com/RxSugar/RxSugar/blob/master/RxSugarTests/UIGestureRecognizer%2BSugarTests.swift) and followed as example.

Is this ok @andersio ?

**Comment by andersio** [comment]
Hmm, yeah. This seems the way so far.

**Comment by andersio** [comment]
https://medium.com/@darjeelingsteve/unit-testing-with-uigesturerecognizer-8342ae1a784e#.srkjfaq3g

This seems relevant too.

**Comment by bitomule** [comment]
I'm not sure I can use that approach. Should I use the Objective-C method in the Swift test? How?

**Comment by andersio** [comment]
Hmm, true. `objc_msgSend` is not available in Swift. Just follow RxSugar's way then. We can find better ways when we extend the extension some point at the future.

**Comment by bitomule** [comment]
Last commit already includes test. Is there anything pending before merge?

File: ReactiveCocoaTests/UIKit/UIGestureRecognizerSpec.swift [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+		}
+	}
+}
+
+private struct TargetActionPair {
+	let target: AnyObject
+	let action: Selector
+}
+
+private class TestTapGestureRecognizer: UITapGestureRecognizer {
```
Comment (andersio) [inline ReactiveCocoaTests/UIKit/UIGestureRecognizerSpec.swift]: Please credit the RxSugar test. 😸

File: ReactiveCocoa/UIKit/UIGestureRecognizer.swift [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import UIKit
+import enum Result.NoError
+
+extension Reactive where Base: UIGestureRecognizer {
+	
+	/// Create a signal which sends a `next` event for each gesture event
+	///
+	/// - returns:
+	///   A trigger signal.
+
```
Comment (andersio) [inline ReactiveCocoa/UIKit/UIGestureRecognizer.swift]: [Nit] Remove the two blank lines around the comment section please.

File: ReactiveCocoa/UIKit/UIGestureRecognizer.swift [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	
+	/// Create a signal which sends a `next` event for each gesture event
+	///
+	/// - returns:
+	///   A trigger signal.
+	
+	public var recognised: Signal<(), NoError> {
+		return Signal { observer in
+			let receiver = CocoaTarget(observer)
+			base.addTarget(receiver, action: #selector(receiver.sendNext))
```
Comment (andersio) [inline ReactiveCocoa/UIKit/UIGestureRecognizer.swift]: Ooooohhh. I thought - with my aged memory in UIKit - the action is just for getting the final ended/recognised event. It turns out to be not. In this case, `trigger` is the right name. I am sorry about this. Would you mind to change it back?

🙏

**Comment by andersio** [comment]
I once thought the action is only for the final `ended` state, and it turns out to be wrong. The name `recognized` in this case would be a little bit off. I am sorry about my mistake.

Anyway, I think we need some bikeshedding with others on this. In practice, we react with different states of the gesture recogniser. So this simple signal is not a meaningful trigger at this stage.

@ReactiveCocoa/reactivecocoa

**Comment by bitomule** [comment]
My idea was just getting back the old "rac_gestureSignal". It doesn't had so many options. Maybe a simple signal is useful for UITapGestureRecognizer but any other GestureRecognizer needs something more complex.

File: ReactiveCocoa/UIKit/UIGestureRecognizer.swift (hunk 4 above)
Comment (NachoSoto) [inline ReactiveCocoa/UIKit/UIGestureRecognizer.swift]: 👍

File: ReactiveCocoa/UIKit/UIGestureRecognizer.swift [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+import enum Result.NoError
+
+extension Reactive where Base: UIGestureRecognizer {
+	
+	/// Create a signal which sends a `next` event for each gesture event
+	///
+	/// - returns:
+	///   A trigger signal.
+	
+	public var recognised: Signal<(), NoError> {
```
Comment (NachoSoto) [inline ReactiveCocoa/UIKit/UIGestureRecognizer.swift]: I would pass the `UIGestureRecognizer` (so `Base`). Users normally need to look at at least `UIGestureRecognizer.state`, for also things like `location`, etc.

File: ReactiveCocoaTests/UIKit/UIGestureRecognizerSpec.swift [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+			gestureRecognizer = TestTapGestureRecognizer()
+			_gestureRecognizer = gestureRecognizer
+		}
+		
+		afterEach {
+			gestureRecognizer = nil
+			expect(_gestureRecognizer).to(beNil())
+		}
+		
+		it("should send a value when the gesture is recognised") {
```
Comment (NachoSoto) [inline ReactiveCocoaTests/UIKit/UIGestureRecognizerSpec.swift]: s/`recognised`/`recognized`

File: ReactiveCocoaTests/UIKit/UIGestureRecognizerSpec.swift [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+			
+			gestureRecognizer.fireGestureEvent( .began)
+			expect(counter) == 2
+			
+			gestureRecognizer.fireGestureEvent(.changed)
+			expect(counter) == 3
+			
+			gestureRecognizer.fireGestureEvent(.ended)
+			expect(counter) == 4
+
```
Comment (NachoSoto) [inline ReactiveCocoaTests/UIKit/UIGestureRecognizerSpec.swift]: Nit: extra space

File: ReactiveCocoaTests/UIKit/UIGestureRecognizerSpec.swift (hunk 2 above)
Comment (NachoSoto) [inline ReactiveCocoaTests/UIKit/UIGestureRecognizerSpec.swift]: This can be `final`.

File: ReactiveCocoa/UIKit/UIGestureRecognizer.swift (hunk 5 above)
Comment (andersio) [inline ReactiveCocoa/UIKit/UIGestureRecognizer.swift]: The name still needs bikeshedding though. `recognised` has an established meaning in the context of `UIGestureRecognizer`.

Comment (andersio) [inline ReactiveCocoa/UIKit/UIGestureRecognizer.swift]: `stateChanged`, hmm?

**Comment by bitomule** [comment]
What name should we use? trigger o recognised?

btw: thanks for your feedback @NachoSoto

**Comment by andersio** [comment]
I propose `stateChanged`, since this doesn't fit what `trigger` is usually meant in RAS.

File: ReactiveCocoa/UIKit/UIGestureRecognizer.swift (hunk 5 above)
Comment (mdiep) [inline ReactiveCocoa/UIKit/UIGestureRecognizer.swift]: This should be `recognized` with a _z_ if the API doesn't change. The project uses US English, but also that's how `UIGestureRecognizer` spells it. :relaxed:

**Comment by bitomule** [comment]
I can't find the way to pass Base as Signal Value. The selector passed to `#Selector`should be `@objc` but my idea was making a Generic CocoaTarget that takes a Generic Value and Generic doesn't work with `@objc`.

Also tried making a UIGestureRecognizerTarget like:

```
internal final class UIGestureRecognizerTarget: NSObject {
	let observer: Observer<UIGestureRecognizer, NoError>
	
	init(_ observer: Observer<UIGestureRecognizer, NoError>) {
		self.observer = observer
	}
	
	@objc func sendNext(value:UIGestureRecognizer) {
		observer.send(value: value)
	}

}
```

But it can't cast `Signal<Base,NoError>` to `Signal<UIGestureRecognizer,NoError>`.

Any ideas?

**Comment by andersio** [comment]
A `CocoaTarget<Value>` that takes a generic closure, e.g. `(Any?) -> Value`, perhaps? Then `sendNext` can be kept to take `Any?` and apply the transform before sending it to the observer.

**Comment by bitomule** [comment]
I've two options. 

First:

```
extension Reactive where Base: UIGestureRecognizer {
	public var recognised: Signal<Base, NoError> {
		return Signal { observer in
			let receiver = UIGestureRecognizerTarget(observer)
			base.addTarget(receiver, action: #selector(receiver.sendNext))
			let disposable = lifetime.ended.observeCompleted(observer.sendCompleted)
			return ActionDisposable { [weak base = self.base] in
				disposable?.dispose()
				base?.removeTarget(receiver, action: #selector(receiver.sendNext))
			}
		}
	}
}
internal final class UIGestureRecognizerTarget<Value>: NSObject {
	let observer: Observer<Value, NoError>
	init(_ observer: Observer<Value, NoError>) {
		self.observer = observer
	}
	@objc func sendNext(value:UIGestureRecognizer) {
		guard let sendValue = value as? Value else {return}
		observer.send(value: sendValue)
	}
}
```

And second one (based on your idea @andersio ):

```
extension Reactive where Base: UIGestureRecognizer {
	public var recognised: Signal<Base?, NoError> {
		return Signal { observer in
			let receiver = UIGestureRecognizerTarget(observer, transform: { gestureRecognizer in
				return gestureRecognizer as? Base
			})
			base.addTarget(receiver, action: #selector(receiver.sendNext))
			let disposable = lifetime.ended.observeCompleted(observer.sendCompleted)
			return ActionDisposable { [weak base = self.base] in
				disposable?.dispose()
				base?.removeTarget(receiver, action: #selector(receiver.sendNext))
			}
		}
	}
}
internal final class UIGestureRecognizerTarget<Value>: NSObject {
	let observer: Observer<Value?, NoError>
	let transform:(Any)->(Value?)
	
	init(_ observer: Observer<Value?, NoError>,transform:@escaping (Any)->(Value?)) {
		self.observer = observer
		self.transform = transform
	}
	
	@objc func sendNext(value:Any) {
		observer.send(value: transform(value))
	}

}
```

I prefer first option maybe with a forced cast (we could be sure that sendNext parameter could be casted to Value)

**Comment by andersio** [comment]
```
internal final class CocoaTarget<Value>: NSObject {
	let observer: Observer<(), NoError>
	let transform: (Any?) -> Value

	init(_ observer: Observer<(), NoError>, transform: @escaping (Any?) -> Value) {
		self.observer = observer
		self.transform = transform
	}

	convenience init(_ observer: Observer<(), NoError>) where Value == Void {
		self.init(observer, transform: { _ in })
	}

	@objc func sendNext(_ receiver: Any?) {
		observer.send(value: transform(receiver))
	}
}
```

The existing user of `CocoaTarget` just needs to parameterize it with `<()>`.

**Comment by bitomule** [comment]
That code doesn't work @andersio . Void isn't a protocol so where cannot be used with Void.

**Comment by bitomule** [comment]
This is the closest I can get to that idea:

```
public var recognised: Signal<Base, NoError> {
		return Signal { observer in
			let receiver = CocoaTarget<Base>(observer, transform: { gestureRecognizer in
				return gestureRecognizer as! Base
			})
			base.addTarget(receiver, action: #selector(receiver.sendNext))
			
			let disposable = lifetime.ended.observeCompleted(observer.sendCompleted)
			
			return ActionDisposable { [weak base = self.base] in
				disposable?.dispose()
				base?.removeTarget(receiver, action: #selector(receiver.sendNext))
			}
		}
	}

internal final class CocoaTarget<Value>: NSObject {
	let observer: Observer<Value, NoError>
	let transform: (Any?) -> Value
	
	init(_ observer: Observer<Value, NoError>, transform: @escaping (Any?) -> Value) {
		self.observer = observer
		self.transform = transform
	}

	
	@objc func sendNext(_ receiver: Any?) {
		observer.send(value: transform(receiver))
	}
}
```

**Comment by andersio** [comment]
That's fine. You can add a `CocoaTargetProtocol` to provide that convenience initializer. We can remove it when concrete type requirements land in Swift 3.1.

**Comment by bitomule** [comment]
Tried this but got Segmentation fault. Swift doesn't love Void.

```
extension CocoaTargetProtocol where Self:CocoaTarget<Void>{
	init(_ observer: Observer<(), NoError>){
		self.init(observer, transform: { _ in })
	}
}
```

Was that your idea?

**Comment by andersio** [comment]
`where Value == Void`. Use associated type constraints.

**Comment by bitomule** [comment]
I'm sorry, I can't understand what you mean. Tried this:

```
internal final class CocoaTarget<Value>: NSObject {
	let observer: Observer<Value, NoError>
	let transform: (Any?) -> Value
	
	init(_ observer: Observer<Value, NoError>, transform: @escaping (Any?) -> Value) {
		self.observer = observer
		self.transform = transform
	}

	
	@objc func sendNext(_ receiver: Any?) {
		observer.send(value: transform(receiver))
	}
}

protocol CocoaTargetProtocol{}
extension CocoaTarget:CocoaTargetProtocol{}

extension CocoaTargetProtocol{
	init<Value>(_ observer: Observer<Value, NoError>) where Value == Void{
		self.init(observer, transform: { _ in })
	}
}
```

And doesn't work. It would be great if you could do a real example of what you mean. If there's a way to do it in Swift 3 I can't find it :(

**Comment by andersio** [comment]
```
protocol CocoaTargetProtocol: class {
    associatedtype Value
    init(_ observer: Observer<Value, NoError>, transform: @escaping (Any?) -> Value)
}

extension CocoaTargetProtocol where Value == Void {
    init<Value>(_ observer: Observer<Value, NoError>) {
        self.init(observer, transform: { _ in })
    }
}
```

**Comment by bitomule** [comment]
Thanks @andersio ,

I think it's ready now. Added 2 more tests to check that signal passed the right gesture recognizer instance and also that state matches the fired event.

I tried to apply all the changes requested :)

File: ReactiveCocoa/CocoaTarget.swift [hunk 8]
```diff
@@ -3,14 +3,29 @@ import ReactiveSwift
...
-		observer.send(value: ())
+		observer.send(value: transform(receiver))
+	}
+}
+
+protocol CocoaTargetProtocol: class {
+	associatedtype Value
+	init(_ observer: Observer<Value, NoError>, transform: @escaping (Any?) -> Value)
+}
+extension CocoaTarget:CocoaTargetProtocol{}
```
Comment (andersio) [inline ReactiveCocoa/CocoaTarget.swift]: A space after `Protocol`, please.

File: ReactiveCocoa/CocoaTarget.swift [hunk 9]
```diff
@@ -3,14 +3,29 @@ import ReactiveSwift
...
 	}
 
+	
 	@objc func sendNext(_ receiver: Any?) {
-		observer.send(value: ())
+		observer.send(value: transform(receiver))
+	}
+}
+
+protocol CocoaTargetProtocol: class {
```
Comment (andersio) [inline ReactiveCocoa/CocoaTarget.swift]: Please mark this and the extension method below with `internal`. 😸

File: ReactiveCocoa/UIKit/UIGestureRecognizer.swift (hunk 3 above)
Comment (andersio) [inline ReactiveCocoa/UIKit/UIGestureRecognizer.swift]: Bump.

File: ReactiveCocoaTests/UIKit/UIGestureRecognizerSpec.swift [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+		
+		it("should send it's gesture recognizer with the fired state") {
+			let signal = gestureRecognizer.reactive.stateChanged
+			weak var signalGestureRecognizer:TestTapGestureRecognizer?
+			signal.observeValues{ recognizer in
+				signalGestureRecognizer = recognizer
+			}
+			
+			gestureRecognizer.fireGestureEvent(.possible)
+			expect(signalGestureRecognizer?.state) == .possible
```
Comment (andersio) [inline ReactiveCocoaTests/UIKit/UIGestureRecognizerSpec.swift]: 👍

**Comment by bitomule** [comment]
Changes done

File: ReactiveCocoa/CocoaTarget.swift [hunk 11]
```diff
@@ -3,14 +3,29 @@ import ReactiveSwift
...
+	associatedtype Value
+	init(_ observer: Observer<Value, NoError>, transform: @escaping (Any?) -> Value)
+}
+extension CocoaTarget:CocoaTargetProtocol{}
+
+extension CocoaTargetProtocol where Value == Void {
+	init(_ observer: Observer<(), NoError>) {
```
Comment (andersio) [inline ReactiveCocoa/CocoaTarget.swift]: Internal here too. 😸

File: ReactiveCocoa/CocoaTarget.swift [hunk 12]
```diff
@@ -3,14 +3,30 @@ import ReactiveSwift
 import enum Result.NoError
 
 /// A target that accepts action messages.
-internal final class CocoaTarget: NSObject {
-	let observer: Observer<(), NoError>
-
-	init(_ observer: Observer<(), NoError>) {
+internal final class CocoaTarget<Value>: NSObject {
+	let observer: Observer<Value, NoError>
```
Comment (andersio) [inline ReactiveCocoa/CocoaTarget.swift]: `private` for these properties.

File: ReactiveCocoa/CocoaTarget.swift [hunk 13]
```diff
@@ -3,14 +3,30 @@ import ReactiveSwift
...
+	let transform: (Any?) -> Value
+	
+	init(_ observer: Observer<Value, NoError>, transform: @escaping (Any?) -> Value) {
```
Comment (andersio) [inline ReactiveCocoa/CocoaTarget.swift]: `internal`.

File: ReactiveCocoa/CocoaTarget.swift [hunk 14]
```diff
@@ -3,14 +3,30 @@ import ReactiveSwift
...
 		self.observer = observer
+		self.transform = transform
 	}
 
+	
 	@objc func sendNext(_ receiver: Any?) {
```
Comment (andersio) [inline ReactiveCocoa/CocoaTarget.swift]: `internal` too.

File: ReactiveCocoa/CocoaTarget.swift [hunk 15]
```diff
@@ -3,14 +3,30 @@ import ReactiveSwift
...
+
+internal protocol CocoaTargetProtocol: class {
+	associatedtype Value
+	init(_ observer: Observer<Value, NoError>, transform: @escaping (Any?) -> Value)
+}
+
+extension CocoaTarget:CocoaTargetProtocol{}
+
+extension CocoaTargetProtocol where Value == Void {
+	init(_ observer: Observer<(), NoError>) {
```
Comment (andersio) [inline ReactiveCocoa/CocoaTarget.swift]: `internal`.

File: ReactiveCocoa/UIKit/UIGestureRecognizer.swift [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+import enum Result.NoError
+
+extension Reactive where Base: UIGestureRecognizer {
+	/// Create a signal which sends a `next` event for each gesture event
+	///
+	/// - returns:
+	///   A trigger signal.
+	public var stateChanged: Signal<Base, NoError> {
+		return Signal { observer in
+			let receiver = CocoaTarget<Base>(observer, transform: { gestureRecognizer in
```
Comment (andersio) [inline ReactiveCocoa/UIKit/UIGestureRecognizer.swift]: You can use trailing closure here. 😸 
_Thought I have commented this before, but can't find it somehow, huh._

File: ReactiveCocoaTests/UIKit/UIGestureRecognizerSpec.swift [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+	}
+}
+
+private struct TargetActionPair {
+	let target: AnyObject
+	let action: Selector
+}
+
+final private class TestTapGestureRecognizer: UITapGestureRecognizer {
+	var targetActionPair: TargetActionPair?
```
Comment (andersio) [inline ReactiveCocoaTests/UIKit/UIGestureRecognizerSpec.swift]: `private` and `fileprivate` as appropriate, please. 😸

File: ReactiveCocoa/CocoaTarget.swift (hunk 8 above)
Comment (andersio) [inline ReactiveCocoa/CocoaTarget.swift]: [Bump] `extension CocoaTarget: CocoaTargetProtocol {}`

File: ReactiveCocoa/CocoaTarget.swift (hunk 11 above)
Comment (andersio) [inline ReactiveCocoa/CocoaTarget.swift]: `internal` please.

**Comment by bitomule** [comment]
@andersio changes done, if you find more things to change please report, I'll change all asap :)

**Comment by andersio** [comment]
@bitomule Please merge https://github.com/bitomule/ReactiveCocoa/pull/5/files and we are good to go. 👍

**Comment by bitomule** [comment]
Merged @andersio

File: ReactiveCocoa/CocoaTarget.swift [hunk 18]
```diff
@@ -3,14 +3,29 @@ import ReactiveSwift
...
-	let observer: Observer<(), NoError>
-
-	init(_ observer: Observer<(), NoError>) {
+internal final class CocoaTarget<Value>: NSObject {
+	private let observer: Observer<Value, NoError>
+	private let transform: (Any?) -> Value
+	
+	internal init(_ observer: Observer<Value, NoError>, transform: @escaping (Any?) -> Value) {
 		self.observer = observer
+		self.transform = transform
```
Comment (NachoSoto) [inline ReactiveCocoa/CocoaTarget.swift]: What take a `transform`? Wouldn't this always be the same implementation to cast to `Value`?

File: ReactiveCocoaTests/UIKit/UIGestureRecognizerSpec.swift [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+					counter += 1
+				}
+			}
+			gestureRecognizer.fireGestureEvent( .began)
+			expect(counter) == 1
+		}
+		
+		it("should send it's gesture recognizer with the fired state") {
+			let signal = gestureRecognizer.reactive.stateChanged
+			weak var signalGestureRecognizer:TestTapGestureRecognizer?
```
Comment (NachoSoto) [inline ReactiveCocoaTests/UIKit/UIGestureRecognizerSpec.swift]: Space after `:`

File: ReactiveCocoaTests/UIKit/UIGestureRecognizerSpec.swift [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+			expect(counter) == 3
+			
+			gestureRecognizer.fireGestureEvent(.ended)
+			expect(counter) == 4
+		}
+		
+		it("should send it's gesture recognizer in signal") {
+			let signal = gestureRecognizer.reactive.stateChanged
+			var counter = 0
+			signal.observeValues{ signalGestureRecognizer in
```
Comment (NachoSoto) [inline ReactiveCocoaTests/UIKit/UIGestureRecognizerSpec.swift]: Space before `{`

File: ReactiveCocoa/UIKit/UIGestureRecognizer.swift [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+import enum Result.NoError
+
+extension Reactive where Base: UIGestureRecognizer {
+	/// Create a signal which sends a `next` event for each gesture event
+	///
+	/// - returns:
+	///   A trigger signal.
+	public var stateChanged: Signal<Base, NoError> {
+		return Signal { observer in
+			let receiver = CocoaTarget<Base>(observer){ gestureRecognizer in
```
Comment (NachoSoto) [inline ReactiveCocoa/UIKit/UIGestureRecognizer.swift]: Space before `{`

File: ReactiveCocoa/CocoaTarget.swift (hunk 18 above)
Comment (andersio) [inline ReactiveCocoa/CocoaTarget.swift]: So this was made to share the implementation by letting the user decide how the `Any?` sender is transformed to feed the observer. Existing users of `CocoaTarget` need not pass through the sender, so it was `Any? -> Void`, while in this case it is `Any? -> UIGestureRecognizer`.

File: ReactiveCocoa/AppKit/NSControl.swift [hunk 22]
```diff
@@ -96,7 +96,7 @@ extension Reactive where Base: NSControl {
 		return associatedValue { base in
 			let (signal, observer) = Signal<(), NoError>.pipe()
 
-			let receiver = CocoaTarget(observer)
+			let receiver = CocoaTarget<()>(observer)
```
Comment (andersio) [inline ReactiveCocoa/AppKit/NSControl.swift]: Sorry for a late nit: I think the type inferrer should be able to infer it. Can you try to remove `<()>`? If it works, please do the same to the one in UIControl too. Thanks.

Comment (bitomule) [inline ReactiveCocoa/AppKit/NSControl.swift]: Worked. Pushed @NachoSoto style changes too.

**Comment by bitomule** [comment]
It's all ready now @andersio ?

**Comment by andersio** [comment]
@bitomule The branch conflict needs to be handled.

**Comment by andersio** [comment]
LGTM. Now let's wait for the tests.

**Comment by andersio** [comment]
LGTM. Now let's wait for the tests.
@mdiep Wanna have a look?

**Pull request merged**